lib/Factory.ts

```diff
@@ -94,4 +95,182 @@ export default class Factory
+            case types.ZERO_OR_MORE_PATH:
+                const zom: A.ZeroOrMorePath = <A.ZeroOrMorePath> result;
+                return this.createZeroOrMorePath(recurse(zom.path));
+            case types.ZERO_OR_ONE_PATH:
+                const zoo: A.ZeroOrOnePath = <A.ZeroOrOnePath> result;
+                return this.createZeroOrOnePath(recurse(zoo.path));
+            default: throw new Error('Unknown Operation type ' + result.type);
+        }
+    }
+
+    /**
+     * Similar to the {@link recurse} function but specifically for expressions.
+     * Both functions call each other while copying.
+     * Should not be called directly since it does not execute the callbacks, these happen in {@link recurse}.
+     * @param {Expression} expr - The Operation to recurse on.
+     * @param { [type: string]: (op: Operation) => RecurseResult } callbacks - A map of required callback Operations.
+     * @returns {Operation} - The copied result.
+     */
+    private recurseExpression(expr: A.Expression, callbacks: { [type: string]: (op: A.Operation) => RecurseResult }): A.Expression
+    {
+        let recurse = (op: A.Operation) => this.recurse(op, callbacks);
+
+        switch(expr.expressionType)
+        {
+            case expressionTypes.AGGREGATE:
+                if (expr.variable)
+                {
+                    const bound = <A.BoundAggregate> expr;
+                    return this.createBoundAggregate(bound.variable, bound.aggregator, <Expression> recurse(bound.expression), bound.distinct, bound.separator);
+                }
+                const aggregate = <A.AggregateExpression> expr;
+                return this.createAggregateExpression(aggregate.aggregator, <Expression> recurse(aggregate.expression), aggregate.distinct, aggregate.separator);
+            case expressionTypes.EXISTENCE:
+                const exist = <A.ExistenceExpression> expr;
+                return this.createExistenceExpression(exist.not, recurse(exist.input));
+            case expressionTypes.NAMED:
+                const named = <A.NamedExpression> expr;
+                return this.createNamedExpression(named.name, <A.Expression[]> named.args.map(recurse));
+            case expressionTypes.OPERATOR:
+                const op = <A.OperatorExpression> expr;
+                return this.createOperatorExpression(op.operator, <A.Expression[]> op.args.map(recurse));
+            case expressionTypes.TERM:
+                const term = <A.TermExpression> expr;
+                return this.createTermExpression(term.term);
+            default: throw new Error('Unknown Expression type ' + expr.expressionType);
+        }
+    }
+}
+
+/**
+ * @interface RecurseResult
+ * @property {Operation} result - The resulting A.Operation.
+ * @property {boolean} recurse - Whether to continue with recursion.
+ */
+export interface RecurseResult
+{
+    result: A.Operation;
```

Comment:
Can this be falsy or not?

Reply:
It can be null, but I can't think of any situation where you would want it to be since it will result in invalid Algebra. So unless you have a use case in mind I want to keep it like this :)

Reply:
Aaah, nvm, I misunderstood.

I thought the callback was going to be called for every operation type, which is of course not true. It will only be called for the callbacks that are defined in the map. And it that case the user has to provide a new operator.

Reply:
Ah yes, that is how it works. All the other ones that don't need to change can be ignored.